runtime/intern.c

```diff
@@ -326,16 +397,26 @@ static void stack_push_items(struct intern_stack* s, value dest, int n) {
   }
 }
 
-static void intern_cleanup(struct intern_stack* s)
+static void intern_cleanup(struct caml_intern_state* s)
 {
-  if (intern_input != NULL) {
-    caml_stat_free (intern_input);
-    intern_input = NULL;
+  s->intern_src = NULL;
+  if (s->intern_input != NULL) {
+    caml_stat_free (s->intern_input);
```

Comment:
I'm a bit worried this isn't always going to work as in stock OCaml. 

Take the following calls:
 - `caml_input_value_from_malloc` with a buffer passed in with `data` 
 - this will call `intern_cleanup(s)` in `input_val_from_block`
 - which will result in the passed `char * data` having `caml_stat_free` called on it
 
 Have I missed something or is that possible?
 I'm wondering if we need a more explicit way of saying: call this free function on the buffer when you clean it up.

Reply:
I think freeing the input data is precisely what we want for `caml_input_value_from_malloc`. See https://github.com/ocaml/ocaml/blob/trunk/runtime/caml/intext.h#L135-L139. In particular, if `s->input_data` is not NULL, then it is supposed to be freed. See https://github.com/ocaml/ocaml/blob/trunk/runtime/intern.c#L44-L46. I believe we match the stock semantics. Am I missing something?

Reply:
Apologies, looks like I got the wrong end of the stick here.